[BEAM-2963] Stage pipeline in DataflowRunner

kennknowles opened the pull request:
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

I've confirmed that the pipeline is uploaded and the metadata set to the actual path, which has the hash appended to avoid collision.

R: @herohde

kennknowles commented:
All failures are caused by mocking in the test. I will replace them with a fake stager to remove this erroneous situation.

herohde commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:
```diff
@@ -516,6 +519,25 @@ public DataflowPipelineJob run(Pipeline pipeline) {
+    RunnerApi.Pipeline protoPipeline = PipelineTranslation.toProto(pipeline);
+    File serializedProtoPipeline;
+    try {
+      serializedProtoPipeline = File.createTempFile("pipeline", ".bin");
```
How about ".pb" as the extention? ".bin" is rather generic.

herohde commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:
```diff
@@ -560,6 +582,7 @@ public DataflowPipelineJob run(Pipeline pipeline) {
     String workerHarnessContainerImage = getContainerImageForJob(options);
     for (WorkerPool workerPool : newJob.getEnvironment().getWorkerPools()) {
       workerPool.setWorkerHarnessContainerImage(workerHarnessContainerImage);
+      workerPool.setMetadata(ImmutableMap.of("pipeline", stagedPipeline.getLocation()));
```
Maybe "pipeline_url" instead?

herohde commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:
```diff
@@ -516,6 +519,25 @@ public DataflowPipelineJob run(Pipeline pipeline) {
+    try (PackageUtil packageUtil = PackageUtil.withDefaultThreadPool()) {
+      stagedPipeline =
+          packageUtil
+              .stageClasspathElements(
```
How are we ensuring that we don't conflict the pipeline URL across jobs or other staged files? Do we do rely on a local temp file name and the MD5 suffix being added implicitly?

herohde commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/Stager.java:
```diff
@@ -25,5 +25,8 @@
  */
 public interface Stager {
   /* Stage files and return a list of packages. */
```
Maybe expand the comment to explain what files are staged by default?

kennknowles commented:
R: @lukecwik as committer

kennknowles commented:
run java precommit

lukecwik commented:
PipelineOptions is our injection mechanism and commonly used within tests/code to override how things work. Attempting to only move out the single parameter would be inconsistent with how the majority of the system passes around information so it would be best if it was left alone or the majority was modified to perform injection another way.

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:
```diff
@@ -514,8 +524,24 @@ public DataflowPipelineJob run(Pipeline pipeline) {
+    List<DataflowPackage> packages = options.getStager().stageDefaultFiles();
+
+    RunnerApi.Pipeline protoPipeline = PipelineTranslation.toProto(pipeline);
+    File serializedProtoPipeline;
```
Also, this change assumes we have write access to a temporary directory on the users system. We currently require read only access to be able to submit a Dataflow pipeline, do you think we should be able to use FileSystems to write to GCS location directly and not use the stager class directly?

kennknowles replied:
It is a good point. I'll see how much I have to reimplement to get that going.

kennknowles replied:
Done

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:
```diff
@@ -64,9 +64,13 @@
+ */
 @Internal
-class PackageUtil implements Closeable {
+public class PackageUtil implements Closeable {
```
We should not make this class public or its methods.

kennknowles replied:
Hmm, yes I first when directly to `PackageUtil` from the `DataflowRunner` but reverted that and used the stager directly. I'll revert this, too, though it doesn't matter since this whole directory shouldn't exist and is just an internal implementation detail.

lukecwik replied:
This still seems to be public

kennknowles replied:
Whoops. Thought I had simply dropped the commit but apparently botched that maneuver. Really done now.

kennknowles commented:
We already have `DataflowRunner.fromOptions(...)`. Hanging on to the options was a design mistake. The runner should not depend on unbounded global (and mutable!) state, but should pull out just the pieces it needs for instantiation. I didn't do it here anyhow.

lukecwik commented on runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java:
```diff
@@ -189,6 +209,30 @@ public SeekableByteChannel answer(InvocationOnMock invocation) throws Throwable
     when(mockGcsUtil.bucketAccessible(GcsPath.fromUri(VALID_PROFILE_BUCKET))).thenReturn(true);
```
`
     when(mockGcsUtil.bucketAccessible(GcsPath.fromUri(VALID_STAGING_BUCKET))).thenReturn(true);
     when(mockGcsUtil.bucketAccessible(GcsPath.fromUri(VALID_STAGING_BUCKET))).thenReturn(true);
`
don't need the duplicate lines, last specified return is always returned on multiple calls.

kennknowles replied:
Didn't touch these lines but I'll dedupe.

lukecwik replied:
Dedupe disappeared?

kennknowles replied:
Hmm, signs point to rebase fail. Done again.

lukecwik commented:
We have avoided using dependency injection and copying forward all dependencies including our dependencies dependencies and so forth is a pain and having many context objects or a god object is a pain and not much better for scoping other then PipelineOptions. Best suggestion (without any implementation traction) was to add support for PipelineOptions to become "immutable".

kennknowles commented:
In this case, there is no pain. We don't need to propagate dependency's dependencies. It is straightforward parameterization.

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:
```diff
@@ -217,69 +225,99 @@ private StagingResult tryStagePackageWithRetry(
+        ByteSource.wrap(attributes.getBytes()).copyTo(Channels.newOutputStream(writer));
+      } else {
+        File sourceFile = attributes.getSource();
+        checkState(sourceFile != null, "Attempt to stage to without file or bytes");
```
awkward sentence: Attempt to stage to without file or bytes

kennknowles replied:
Rephrased.

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:
```diff
@@ -456,7 +477,36 @@ public static PackageAttributes forFileToStage(File source, String stagingPath)
+      try (CountingOutputStream countingOutputStream = new CountingOutputStream(hashStream)) {
+        ByteSource.wrap(bytes).copyTo(countingOutputStream);
+        countingOutputStream.flush();
+        size = countingOutputStream.getCount();
```
Why is size not just the length of the bytes array?

In the above method it looks like we just used the same methodology for files and for directories to count their size instead of short circuiting and using the length of the file as the size.

kennknowles replied:
Done

kennknowles commented:
PTAL. Added an assertion that every valid pipeline has this field set. The assertion is already used by many test cases. Further coverage is most appropriate via integration testing once this field is in use, or by the improved injection and faking I mentioned, either one would be beyond this PR. Added a commit to bump the Kafka version to get test signal. It should be resolved shortly by #4008 or similar and I will drop that commit before merging.

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/Stager.java:
```diff
@@ -20,10 +20,32 @@
+  List<DataflowPackage> stageFiles(List<String> filesToStage);
+
+  /**
+   * Stage bytes to a target file name wherever this stager stages things.
```
"target" implies the full output path instead of just a local name for the bytes which the stager can use to place somewhere. Consider using a different name then "target"

kennknowles replied:
Done. Tried `baseName` on for size.

lukecwik commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:
```diff
@@ -456,7 +477,36 @@ public static PackageAttributes forFileToStage(File source, String stagingPath)
+        {
+      long size;
+      String hash;
+      Hasher hasher = Hashing.md5().newHasher();
```
hasher.putBytes(bytes)? instead of wrapping in a funnel/outputstream/...

kennknowles replied:
Done. (was copy pasta from file processing, to make it easy to see that it is identical)
